src/libos/crates/host-socket/src/stream/states/connected/send.rs

```diff
@@ -168,19 +191,76 @@ impl<A: Addr + 'static, R: Runtime> ConnectedStream<A, R> {
             if !inner.send_buf.is_empty() {
                 stream.do_send(&mut inner);
             } else if inner.is_shutdown == ShutdownStatus::PreShutdown {
+                // The buffer is empty and the write side is shutdown by the user. We can safely shutdown host file here.
+                let _ = stream.common.host_shutdown(Shutdown::Write);
                 inner.is_shutdown = ShutdownStatus::PostShutdown
             }
         };
 
         // Generate the async send request
         let msghdr_ptr = inner.new_send_req();
 
+        trace!("send submit request");
         // Submit the async send to io_uring
         let io_uring = self.common.io_uring();
         let host_fd = Fd(self.common.host_fd() as _);
         let handle = unsafe { io_uring.sendmsg(host_fd, msghdr_ptr, 0, complete_fn) };
         inner.io_handle.replace(handle);
     }
+
+    pub fn cancel_send_requests(&self) {
+        let io_uring = self.common.io_uring();
+        let inner = self.sender.inner.lock().unwrap();
+        if let Some(io_handle) = &inner.io_handle {
+            unsafe { io_uring.cancel(io_handle) };
+        }
+    }
+
+    // Normally, We will always try to send as long as the kernel send buf is not empty. However, if the user calls close, we will wait LINGER time
+    // and then cancel on-going or new-issued send requests.
+    pub async fn try_empty_send_buf_when_close(&self) {
+        let inner = self.sender.inner.lock().unwrap();
+        debug_assert!(inner.is_shutdown());
+        if inner.send_buf.is_empty() {
+            return;
+        }
+
+        // Wait for linger time to empty the kernel buffer or cancel subsequent requests.
+        drop(inner);
+        const DEFUALT_LINGER_TIME: usize = 10;
+        let poller = Poller::new();
+        let mask = Events::ERR | Events::IN;
+        self.common.pollee().connect_poller(mask, &poller);
+
+        loop {
+            let pending_request_exist = {
+                let inner = self.sender.inner.lock().unwrap();
+                inner.io_handle.is_some()
+            };
+
+            if pending_request_exist {
+                let mut timeout = Some(Duration::from_secs(DEFUALT_LINGER_TIME as u64));
```

Comment:
Each wait timeout is 10 seconds?

Reply:
Yes. Timeout 10 seconds for each request. However, send can rarely block. And the cancel request should take effect in a snap. Thus, this function shouldn't take a long time.